storage/namespace.go

```diff
@@ -443,14 +443,53 @@ func (n *dbNamespace) Flush(
 		}
 	}
 
-	// if nil, succeeded
+	res := multiErr.FinalError()
+	n.metrics.flush.ReportSuccessOrError(res, n.nowFn().Sub(callStart))
+	return res
+}
 
-	if res := multiErr.FinalError(); res != nil {
-		n.metrics.flush.ReportError(n.nowFn().Sub(callStart))
-		return res
+func (n *dbNamespace) FlushState(blockStart time.Time) fileOpState {
+	// NB(r): The logic below is not extensively clear just by reading the
+	// code. Essentially if all are success, we return success, if any
+	// are failed then we return failed with the minimum num failures to
+	// give all shards the chance of max retries - otherwise if any in progress
+	// we return in progress and if any is not started we return not started.
+	// This is implemented the way it is to avoid iterating over all shards
+	// and reading/querying their states multiple times.
+	anyFailed := false
+	minNumFailures := int(math.MaxInt32)
+
+	n.RLock()
+	defer n.RUnlock()
+
+	for _, shard := range n.shards {
+		if shard == nil {
+			continue
+		}
+		state := shard.FlushState(blockStart)
+		status := state.Status
+		if status == fileOpNotStarted || status == fileOpInProgress {
+			// Any not started or in progress overrides other states
+			return state
+		}
+		if status == fileOpFailed {
+			anyFailed = true
+			if state.NumFailures < minNumFailures {
```

Comment:
Seems like if we make the decision whether we should flush here in namespace.go as opposed to passing back the state to the caller, should make it less confusing?